### ISSUE TITLE
Redundant memref.alloca and copies are created in hoisting padding path

### ISSUE THREAD

**Issue by hanhanW** [google/iree#9016]
I'm prototyping using hoist-padding in IREE, and found that there are redundant memref.alloca ops and copies. Input IR for bufferization: https://gist.githubusercontent.com/hanhanW/b623547c0572619cfa5730a2fe6b2633/raw

The below is a snippet about hoisted padding.

```mlir
  %8 = linalg.init_tensor [2, 96, 16, 32] : tensor<2x96x16x32xf32>
  %9 = tensor.cast %8 : tensor<2x96x16x32xf32> to tensor<?x?x16x32xf32>
...
      %16 = scf.for %arg2 = %c0 to %c64 step %c32 iter_args(%arg3 = %9) -> (tensor<?x?x16x32xf32>) {
        %18 = affine.apply affine_map<(d0) -> (d0 ceildiv 32)>(%arg2)
        %19 = scf.for %arg4 = %c0 to %c1536 step %c16 iter_args(%arg5 = %arg3) -> (tensor<?x?x16x32xf32>) {
          %20 = affine.apply affine_map<(d0) -> (d0 ceildiv 16)>(%arg4)
          %21 = vector.transfer_read %15[%arg4, %arg2], %cst {in_bounds = [true, true]} : tensor<1536x64xf32>, vector<16x32xf32>
          %22 = vector.transfer_write %21, %arg5[%18, %20, %c0, %c0] {in_bounds = [true, true]} : vector<16x32xf32>, tensor<?x?x16x32xf32>
          scf.yield %22 : tensor<?x?x16x32xf32>
        }
        scf.yield %19 : tensor<?x?x16x32xf32>
      }
```

After bufferization and some cleanup, there are two `memref<2x96x16x32xf32>` and one redundant copy.

Full IR: https://gist.githubusercontent.com/hanhanW/0fed1063bdf8a4b2b33a140f9f3ced37/raw

snippet:

```mlir
  %12 = memref.alloca() {alignment = 128 : i64} : memref<2x96x16x32xf32>
  %13 = memref.alloca() {alignment = 128 : i64} : memref<2x96x16x32xf32>
  %14 = memref.alloca() {alignment = 128 : i64} : memref<96x12x16xf32>
  %15 = memref.alloca() {alignment = 128 : i64} : memref<96x12x16xf32>
  scf.for %arg0 = %8 to %c64 step %9 {
    %16 = memref.subview %0[%arg0, 0] [64, 1536] [1, 1] : memref<64x1536xf32> to memref<64x1536xf32, affine_map<(d0, d1)[s0] -> (d0 * 1536 + s0 + d1)>>
    scf.for %arg1 = %10 to %c64 step %11 {
      %17 = memref.subview %6[%arg0, %arg1] [64, 64] [1, 1] : memref<64x64xf32> to memref<64x64xf32, affine_map<(d0, d1)[s0] -> (d0 * 64 + s0 + d1)>>
      %18 = memref.subview %4[%arg1] [64] [1] : memref<64xf32> to memref<64xf32, affine_map<(d0)[s0] -> (d0 + s0)>>
      %19 = memref.subview %2[0, %arg1] [1536, 64] [1, 1] : memref<1536x64xf32> to memref<1536x64xf32, affine_map<(d0, d1)[s0] -> (d0 * 64 + s0 + d1)>>
      linalg.generic {indexing_maps = [affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>, affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%12 : memref<2x96x16x32xf32>) outs(%13 : memref<2x96x16x32xf32>) {
      ^bb0(%arg2: f32, %arg3: f32):
        linalg.yield %arg2 : f32
      }
      scf.for %arg2 = %c0 to %c64 step %c32 {
        %20 = affine.apply affine_map<(d0) -> (d0 ceildiv 32)>(%arg2)
        scf.for %arg3 = %c0 to %c1536 step %c16 {
          %21 = affine.apply affine_map<(d0) -> (d0 ceildiv 16)>(%arg3)
          %22 = vector.transfer_read %19[%arg3, %arg2], %cst {in_bounds = [true, true]} : memref<1536x64xf32, affine_map<(d0, d1)[s0] -> (d0 * 64 + s0 + d1)>>, vector<16x32xf32>
          vector.transfer_write %22, %13[%20, %21, %c0, %c0] {in_bounds = [true, true]} : vector<16x32xf32>, memref<2x96x16x32xf32>
        }
      }
```

`%12` is redundant, and we don't have to copy from `%12` to `%13`.

**Comment by hanhanW** [comment]
This can be reproduced with https://github.com/google/iree/pull/9015#issuecomment-1111693583 or using `iree-opt` with bufferization transforms.

**Comment by hanhanW** [comment]
@matthias-springer -- just wanna check in if this is on your radar. Do you need more context or need me to simplify the IR?

**Comment by matthias-springer** [comment]
I think D124849 should solve the issue. Can you verify?

**Comment by hanhanW** [comment]
oops, I did not notice that there are holidays in EU. I'll verify it tomorrow, thanks for the quick fix!! :D

**Comment by hanhanW** [comment]
Yes, it does address the issue, thanks again!

**Comment by hanhanW** [comment]
Verified that it's fixed by D124849
